velox/docs/develop/window.rst

```diff
@@ -0,0 +1,202 @@
+The frame numbers increase by 1 for each row.
+
+**RANGE mode**
+
+In RANGE mode, all peer rows have the same frame number.
+Rows are peers if they have the same values for the ORDER BY field.
+A frame start of CURRENT ROW refers to the first peer row of the current row,
+while a frame end of CURRENT ROW refers to the last peer row of the current row.
+If no ORDER BY is specified, all rows are considered peers of the current row.
+
+**Misc**
+
+UNBOUNDED PRECEDING and UNBOUNDED FOLLOWING always refer to the first and
+last rows of the partition in either mode.
+
+
+Window frame indices
+--------------------
+
+As window functions are evaluated for each row, the Window operator provides
+each function a buffer of frame_start and frame_end indices at each
+WindowFunction::apply call.
+
+**Note**: Its possible during calculations that the frame indices are values
+before or after the partition rows. In this case, the frame indices get bound
+to the first and last rows of the partition.
+
+As an example, for frame *ROWS BETWEEEN 2 PRECEDING AND 2 FOLLOWING*, the
+frame indices will be as follows:
+
+.. code-block::
+
+  row_index    partition_col        order_by_col        frame_start         frame_end
+       0            1                     1                   0                 2
+       1            1                     2                   0                 3
+       2            1                     2                   0                 4
+       3            1                     3                   1                 5
+       4            1                     4                   2                 6
+       5            1                     4                   3                 7
+       6            1                     4                   4                 7
+       7            1                     5                   5                 7
+
+On the other hand, for frame *RANGE BETWEEEN 2 PRECEDING AND 2 FOLLOWING*,
+the frame indices will be as follows
+
+.. code-block::
+
+  row_index    partition_col        order_by_col        frame_start         frame_end
+       0            1                     1                   0                 3
+       1            1                     2                   0                 6
+       2            1                     2                   0                 6
+       3            1                     3                   0                 7
+       4            1                     4                   1                 7
+       5            1                     4                   1                 7
+       6            1                     4                   1                 7
+       7            1                     5                   3                 7
+
+Empty frames
+------------
+
+Its possible that certain rows have empty frames. Empty frames occur when :
+
+* Both frameStart and frameEnds fall before the first partition row.
```

Comment:
Should this use frame_start and frame_end for consistency?

Reply:
Yeah, made the usage consistent across the writing.